### PR TITLE
Added Account and Transaction entities, the repository and service levels for them

### PR DESCRIPTION
**Pull request by Vedroid**

### CODE REVIEW

File: src/main/java/ua/vedroid/bankservice/entity/Account.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+@Table(name = "accounts")
+@Data
+public class Account {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @Column(name = "account_number", unique = true)
+    private String accountNumber;
+    private Currency currency;
+    private Double balance;
```
Comment (Mandybur800) [inline src/main/java/ua/vedroid/bankservice/entity/Account.java]: We need BigDecimal

Comment (Mandybur800) [reply src/main/java/ua/vedroid/bankservice/entity/Account.java]: and add some constrains

File: src/main/java/ua/vedroid/bankservice/entity/Transaction.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @ManyToOne
+    @JoinColumn(name = "account_from")
+    private Account fromAccount;
+    @ManyToOne
+    @JoinColumn(name = "account_to")
+    private Account toAccount;
+    private LocalDateTime date;
+    private Double amount;
```
Comment (Mandybur800) [inline src/main/java/ua/vedroid/bankservice/entity/Transaction.java]: same

File: src/main/java/ua/vedroid/bankservice/util/CurrencyConverter.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+            return JsonParser
+                    .parseReader(new InputStreamReader((InputStream) request.getContent()))
+                    .getAsJsonObject()
+                    .get("result")
+                    .getAsDouble();
+        } catch (IOException e) {
+            throw new CurrencyConvertException("Failed to convert currency", e);
+        }
+    }
+}
```
Comment (Mandybur800) [inline src/main/java/ua/vedroid/bankservice/util/CurrencyConverter.java]: just remove this class, because it is part 3

**Review by Mandybur800 — CHANGES_REQUESTED** [review]

File: src/main/java/ua/vedroid/bankservice/entity/Account.java (hunk 1 above)
Comment (Vedroid) [reply src/main/java/ua/vedroid/bankservice/entity/Account.java]: I know that floating point numbers cannot be used for precise calculations. But the assignment says double
image

Comment (boroda4436) [reply src/main/java/ua/vedroid/bankservice/entity/Account.java]: @Vedroid Sorry, our fault. Better use BigDeimal

File: src/main/java/ua/vedroid/bankservice/repository/AccountRepository.java [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+public interface AccountRepository extends JpaRepository<Account, Long> {
+    @Query("FROM Account a LEFT JOIN FETCH a.user u WHERE u.phoneNumber = ?1")
+    List<Account> findAllByUserPhoneNumber(String phoneNumber);
+
+    @Query("FROM Account a LEFT JOIN FETCH a.user WHERE a.accountNumber = ?1")
+    Optional<Account> findByAccountNumber(String accountNumber);
+
+    @Modifying
+    @Transactional
```
Comment (boroda4436) [inline src/main/java/ua/vedroid/bankservice/repository/AccountRepository.java]: @Vedroid we don't need to add this annotation to the repository methods. Better add this annotation on the service layer if you have multiple read/write operations in that method

**Review by boroda4436 — CHANGES_REQUESTED** [review]

File: pom.xml [hunk 5]
```diff
@@ -54,6 +54,11 @@
             <artifactId>spring-boot-starter-test</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>com.google.code.gson</groupId>
+            <artifactId>gson</artifactId>
```
Comment (KseniiaMakarova) [inline pom.xml]: do you use this at this point?

File: src/main/java/ua/vedroid/bankservice/repository/AccountRepository.java [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+import java.util.List;
+import java.util.Optional;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Modifying;
+import org.springframework.data.jpa.repository.Query;
+import ua.vedroid.bankservice.entity.Account;
+
+public interface AccountRepository extends JpaRepository<Account, Long> {
+    @Query("FROM Account a LEFT JOIN FETCH a.user u WHERE u.phoneNumber = ?1")
+    List<Account> findAllByUserPhoneNumber(String phoneNumber);
```
Comment (KseniiaMakarova) [inline src/main/java/ua/vedroid/bankservice/repository/AccountRepository.java:11]: does it work without `@Query` annotation?

File: src/main/java/ua/vedroid/bankservice/repository/AccountRepository.java [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+
+    @Query("FROM Account a LEFT JOIN FETCH a.user WHERE a.accountNumber = ?1")
+    Optional<Account> findByAccountNumber(String accountNumber);
```
Comment (KseniiaMakarova) [inline src/main/java/ua/vedroid/bankservice/repository/AccountRepository.java:14]: same - won't the user be fetched by default here?

File: src/main/java/ua/vedroid/bankservice/repository/AccountRepository.java [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+
+    @Modifying
+    @Query("UPDATE Account SET isActive = false WHERE accountNumber = ?1")
+    void blockByAccountNumber(String accountNumber);
```
Comment (KseniiaMakarova) [inline src/main/java/ua/vedroid/bankservice/repository/AccountRepository.java]: let's better set `active` to false somewhere in constructor and then just call `save()`?

File: src/main/java/ua/vedroid/bankservice/repository/TransactionRepository.java [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+
+import java.util.List;
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
+import ua.vedroid.bankservice.entity.Transaction;
+
+public interface TransactionRepository extends JpaRepository<Transaction, Long> {
+    @Query("FROM Transaction t JOIN FETCH t.fromAccount fa JOIN FETCH t.toAccount ta "
+            + "WHERE fa.accountNumber = ?1 OR ta.accountNumber = ?1")
```
Comment (KseniiaMakarova) [inline src/main/java/ua/vedroid/bankservice/repository/TransactionRepository.java:11]: is it obligatory to do join fetch here?

File: src/main/java/ua/vedroid/bankservice/service/AccountService.java [hunk 10]
```diff
@@ -0,0 +1,14 @@
+package ua.vedroid.bankservice.service;
+
+import java.util.List;
+import ua.vedroid.bankservice.entity.Account;
+
+public interface AccountService {
+    Account create(Account account);
```
Comment (KseniiaMakarova) [inline src/main/java/ua/vedroid/bankservice/service/AccountService.java]:
```suggestion
    Account save(Account account);
```

File: src/main/java/ua/vedroid/bankservice/service/impl/TransactionServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+        Account fromAccount = accountService.findByAccountNumber(dto.getFromAccount());
+        Account toAccount = accountService.findByAccountNumber(dto.getToAccount());
+        BigDecimal amount = dto.getAmount();
+        BigDecimal fromAccountBalance = fromAccount.getBalance();
+        BigDecimal currentAmount = fromAccountBalance.subtract(amount);
+        if (currentAmount.compareTo(BigDecimal.valueOf(0)) < 0) {
+            throw new InsufficientFundsException("Insufficient funds on account " + fromAccount);
+        }
+        fromAccount.setBalance(currentAmount);
+        toAccount.setBalance(toAccount.getBalance().add(amount));
```
Comment (KseniiaMakarova) [inline src/main/java/ua/vedroid/bankservice/service/impl/TransactionServiceImpl.java:37]: are you saving the updated balance?

File: src/main/java/ua/vedroid/bankservice/service/impl/TransactionServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+                .type(Transaction.TransactionType.INCOMING)
+                .build();
+        transactionRepository.save(outcomingTransaction);
+        transactionRepository.save(incomingTransaction);
+    }
+
+    @Override
+    public List<Transaction> getHistory(String accountNumber, int page, int limit) {
+        PageRequest pageRequest = PageRequest
+                .of(page - 1, limit, Sort.by("date").descending().and(Sort.by("id")));
```
Comment (KseniiaMakarova) [inline src/main/java/ua/vedroid/bankservice/service/impl/TransactionServiceImpl.java]: why -1?

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: src/main/java/ua/vedroid/bankservice/service/impl/TransactionServiceImpl.java (hunk 12 above)
Comment (Vedroid) [reply src/main/java/ua/vedroid/bankservice/service/impl/TransactionServiceImpl.java]: I wanted Page numbering to start at 1 `/accounts/history/13?page=1`. But it’s probably better not to do it

File: src/main/java/ua/vedroid/bankservice/repository/AccountRepository.java (hunk 6 above)
Comment (Vedroid) [reply src/main/java/ua/vedroid/bankservice/repository/AccountRepository.java:11]: This is how one request to the database happens instead of two.

**Review by KseniiaMakarova — APPROVED** [review]

**Pull request merged**